src/entt/entity/storage.hpp

```diff
@@ -233,25 +307,41 @@ template<typename... Lhs, typename... Rhs>
  * @tparam Allocator Type of allocator used to manage memory and elements.
  */
 template<typename Type, typename Entity, typename Allocator, typename>
-class basic_storage: public basic_sparse_set<Entity, typename std::allocator_traits<Allocator>::template rebind_alloc<Entity>> {
+class basic_storage
+    : public basic_sparse_set<Entity, typename std::allocator_traits<Allocator>::template rebind_alloc<Entity>>,
+      protected Allocator,
+      protected component_traits<Type> {
+    template<typename Container, std::size_t Size>
+    friend struct internal::static_storage_accessor;
+    template<typename Storage>
+    friend struct internal::dynamic_storage_accessor;
+
     using alloc_traits = std::allocator_traits<Allocator>;
     static_assert(std::is_same_v<typename alloc_traits::value_type, Type>, "Invalid value type");
     using container_type = std::vector<typename alloc_traits::pointer, typename alloc_traits::template rebind_alloc<typename alloc_traits::pointer>>;
     using underlying_type = basic_sparse_set<Entity, typename alloc_traits::template rebind_alloc<Entity>>;
     using underlying_iterator = typename underlying_type::basic_iterator;
 
     static constexpr bool is_pinned_type_v = !(std::is_move_constructible_v<Type> && std::is_move_assignable_v<Type>);
+    static constexpr bool is_dynamically_sized_type_v = internal::is_component_trait_dynamically_sized<component_traits<Type>>::value;
 
     [[nodiscard]] auto &element_at(const std::size_t pos) const {
-        return payload[pos / traits_type::page_size][fast_mod(pos, traits_type::page_size)];
+        auto *const page = payload[pos / traits_type::page_size];
+        std::size_t const offset = fast_mod(pos, traits_type::page_size);
+
+        if constexpr(is_dynamically_sized_type_v) {
+            return traits_type::element_at(page, offset);
+        } else {
+            return page[offset];
+        }
     }
 
     auto assure_at_least(const std::size_t pos) {
         const auto idx = pos / traits_type::page_size;
 
         if(!(idx < payload.size())) {
             auto curr = payload.size();
-            allocator_type allocator{get_allocator()};
+            allocator_type &allocator = get_allocator();
```

Comment:
No way you should have a `get_allocator` function returning a reference. Really.

Reply:
See above retreieving the allocator by-value can add overhead for stateful allocators.